File: python/be/contract/chain/sender.py
"""Sending transactions to the local chain -- and refusing to send anywhere else.

Mirror of node/be/contract/chain/sender.js. anvil signs for its unlocked
accounts through ``eth_sendTransaction``; the client never holds a key. The
one guard that matters is the chain id: 31337 or nothing.
"""

from __future__ import annotations

import time

from be.contract.chain import abi
from be.contract.chain.ethcall import ChainUnreachable, RevertError, rpc

ANVIL_CHAIN_ID = 31337


class WrongChain(Exception):
    def __init__(self, actual: int) -> None:
        super().__init__(f"refusing to send: chainId is {actual}, this client only writes to anvil ({ANVIL_CHAIN_ID})")


class Reverted(Exception):
    def __init__(self, decoded: dict, context: str) -> None:
        args = ", ".join(str(a) for a in decoded["args"])
        super().__init__(f"{context} reverted: {decoded['name']}({args})")
        self.error = decoded["name"]
        self.args = decoded["args"]
        self.selector = decoded.get("selector")
        self.raw = decoded.get("raw")


class Sender:
    def __init__(self, url: str, errors: list[str]) -> None:
        self.url = url
        self.error_table = abi.error_table(errors)
        self.verified = False

    def assert_chain(self) -> None:
        if self.verified:
            return
        cid = int(rpc("eth_chainId", [], self.url), 16)
        if cid != ANVIL_CHAIN_ID:
            raise WrongChain(cid)
        self.verified = True

    def call(self, from_: str, to: str, signature: str, args: list, context: str | None = None) -> str:
        data = abi.encode_call(signature, args)
        try:
            return rpc("eth_call", [{"from": from_, "to": to, "data": data}, "latest"], self.url)
        except RevertError as err:
            raise self._translate(err, context or signature) from err

    def send(self, from_: str, to: str, signature: str, args: list, context: str | None = None) -> dict:
        self.assert_chain()
        ctx = context or signature
        data = abi.encode_call(signature, args)
        try:
            rpc("eth_call", [{"from": from_, "to": to, "data": data}, "latest"], self.url)
        except RevertError as err:
            raise self._translate(err, ctx) from err
        tx_hash = rpc("eth_sendTransaction", [{"from": from_, "to": to, "data": data, "gas": hex(3_000_000)}], self.url)
        receipt = self.wait_for_receipt(tx_hash)
        if receipt.get("status") != "0x1":
            try:
                rpc("eth_call", [{"from": from_, "to": to, "data": data}, receipt["blockNumber"]], self.url)
            except RevertError as err:
                raise self._translate(err, ctx + " (mined but failed)") from err
            raise Reverted({"name": "unknown", "args": [], "raw": None}, ctx + " (mined but failed; no reason recoverable)")
        return receipt

    def wait_for_receipt(self, tx_hash: str, attempts: int = 50) -> dict:
        for _ in range(attempts):
            try:
                r = rpc("eth_getTransactionReceipt", [tx_hash], self.url)
            except ChainUnreachable:
                r = None
            if r:
                return r
            time.sleep(0.1)
        raise ChainUnreachable(f"no receipt for {tx_hash} after {attempts} polls")

    def _translate(self, err: RevertError, context: str) -> Reverted:
        decoded = abi.decode_error(err.data or "0x", self.error_table)
        return Reverted(decoded, context)

File: node/be/contract/chain/sender.js
'use strict';

const { rpc, ChainUnreachable } = require('./ethcall');
const abi = require('./abi');

/**
 * Sending transactions to the local chain -- and refusing to send anywhere else.
 *
 * Why no private keys anywhere in this repository
 * ------------------------------------------------
 * anvil starts with ten unlocked accounts and signs on their behalf through
 * `eth_sendTransaction`. That is the whole signing story here: the client
 * names a `from` address, the node signs. No key is read, stored, logged or
 * passed on a command line, and there is nothing to leak.
 *
 * The one guard that matters is the chain: before the first send, the sender
 * asks `eth_chainId` and refuses to continue unless it is 31337. A client
 * pointed at any other RPC by mistake -- a testnet, a fork, a mainnet -- gets
 * WrongChain, not a transaction.
 *
 * Reverts are decoded. A revert on a send is a legitimate answer from a
 * reachable chain, and its custom error is usually the very thing a test is
 * asserting, so it surfaces as `Reverted` with `name` and `args` decoded
 * against the exchange's error table. A transport failure is still
 * ChainUnreachable.
 */

const ANVIL_CHAIN_ID = 31337;

class WrongChain extends Error {
  constructor(actual) {
    super(`refusing to send: chainId is ${actual}, this client only writes to anvil (${ANVIL_CHAIN_ID})`);
    this.name = 'WrongChain';
  }
}

class Reverted extends Error {
  constructor(decoded, context) {
    super(`${context} reverted: ${decoded.name}(${decoded.args.map(String).join(', ')})`);
    this.name = 'Reverted';
    this.error = decoded.name;
    this.args = decoded.args;
    this.selector = decoded.selector;
    this.raw = decoded.raw;
  }
}

class Sender {
  constructor({ url, errors = [] }) {
    this.url = url;
    this.errorTable = abi.errorTable(errors);
    this.verified = false;
  }

  async assertChain() {
    if (this.verified) return;
    const id = Number(BigInt(await rpc('eth_chainId', [], this.url)));
    if (id !== ANVIL_CHAIN_ID) throw new WrongChain(id);
    this.verified = true;
  }

  /** eth_call from `from`, decoding a revert into a Reverted error. */
  async call(from, to, signature, args, context = signature) {
    const data = abi.encodeCall(signature, args);
    try {
      return await rpc('eth_call', [{ from, to, data }, 'latest'], this.url);
    } catch (err) {
      throw this._translate(err, context);
    }
  }

  /**
   * Send and wait for the receipt. Simulates first with eth_call so a revert
   * comes back decoded instead of as a mined failure with no reason.
   */
  async send(from, to, signature, args, context = signature) {
    await this.assertChain();
    const data = abi.encodeCall(signature, args);
    // simulate: anvil returns the revert data on eth_call, not on the receipt
    try {
      await rpc('eth_call', [{ from, to, data }, 'latest'], this.url);
    } catch (err) {
      throw this._translate(err, context);
    }
    const hash = await rpc('eth_sendTransaction', [{ from, to, data, gas: '0x' + (3_000_000).toString(16) }], this.url);
    const receipt = await this.waitForReceipt(hash);
    if (receipt.status !== '0x1') {
      // The simulation passed and the mined transaction did not: the state
      // moved between the two (on anvil, usually the block timestamp). Re-run
      // the call at the mined block to recover the actual reason.
      try {
        await rpc('eth_call', [{ from, to, data }, receipt.blockNumber], this.url);
      } catch (err) {
        throw this._translate(err, context + ' (mined but failed)');
      }
      throw new Reverted({ name: 'unknown', args: [], raw: null }, context + ' (mined but failed; no reason recoverable)');
    }
    return receipt;
  }

  async waitForReceipt(hash, attempts = 50) {
    for (let i = 0; i < attempts; i++) {
      const r = await rpc('eth_getTransactionReceipt', [hash], this.url).catch(() => null);
      if (r) return r;
      await new Promise((res) => setTimeout(res, 100));
    }
    throw new ChainUnreachable('no receipt for ' + hash + ' after ' + attempts + ' polls');
  }

  _translate(err, context) {
    if (err instanceof ChainUnreachable) return err;
    if (!err.reverted) return err;
    const decoded = abi.decodeError(err.data || '0x', this.errorTable);
    return new Reverted(decoded, context);
  }
}

module.exports = { Sender, WrongChain, Reverted, ANVIL_CHAIN_ID };
